### PR TITLE
[Contrib] Captive Portal like lock for internet bound connection on an offline node

### PR DESCRIPTION
**Pull request by darkdrgn2k**
Redirects to Pi's webserver when using WIFI Access Point on a node that will not route ipv4 (Internet) traffic.

- Removed IPV4 masquerading
- Addes redirect to 10.0.0.1 for all port 80 bound traffic
- Adds 404 redirect to / for nginx

### CODE REVIEW

**Comment by darkdrgn2k** [comment]
issue #210 

@DaniellMesquita - Thoughts?

**Comment by darkdrgn2k** [comment]
>  Including an uninstall script,

Its a contrib file. Does it need an uninstall script? 
I made a backup of the nat file though.

File: contrib/offline/redirect-port-80.sh [hunk 1]
```diff
@@ -1,5 +1,10 @@
 #!/bin/sh
 
+# Backup file
+if ! [ -f "/etc/hostapd/nat.sh.org" ]; then
+  sudo cp /etc/hostapd/nat.sh /etc/hostapd/nat.sh.org
```
Comment (benhylau) [inline contrib/offline/redirect-port-80.sh]: I think in other places in the codebase we have been using `.bak` for backup files

File: contrib/offline/redirect-port-80.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+# Backup file
+if ! [ -f "/etc/hostapd/nat.sh.org" ]; then
+  sudo cp /etc/hostapd/nat.sh /etc/hostapd/nat.sh.org
+fi
+
+# Redirect all IPv4 80 traffic to the pi
+echo iptables -t nat -I PREROUTING -i wlan-ap -p tcp --dport 80 -j DNAT --to-destination 10.0.0.1:80 | sudo tee --append  /etc/hostapd/nat.sh > /dev/null 
+
+# Prevent masquarading out ipv4
```
Comment (benhylau) [inline contrib/offline/redirect-port-80.sh]: `masquerading`

Comment (benhylau) [reply contrib/offline/redirect-port-80.sh]: `IPv4` as above

File: contrib/offline/redirect-port-80.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+# This is prevent IPTUNNEL and routing to the internet (Exit node)
```
Comment (benhylau) [inline contrib/offline/redirect-port-80.sh]: `to` prevent

File: contrib/offline/redirect-port-80.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+sudo sed -i "/iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE/d" /etc/hostapd/nat.sh 
+
+# Set nginx to redirect any 404 errors to /
+sudo sed -i "s/}/    error_page 404 =200 \/index.html;\\n}/" /etc/nginx/sites-enabled/main.conf
```
Comment (benhylau) [inline contrib/offline/redirect-port-80.sh]: This replaces `}`? Is it safe?

**Comment by benhylau** [comment]
@darkdrgn2k @makeworld-the-better-one contribs generally don't have uninstalls

File: contrib/offline/redirect-port-80.sh (hunk 4 above)
Comment (darkdrgn2k) [reply contrib/offline/redirect-port-80.sh]: hacky I know
in this. file there should be only one}

don't know how else to do this Esp in bash

@benhylau any ideas? how to insert a line on the Second last line of a file

Comment (benhylau) [reply contrib/offline/redirect-port-80.sh]: `sed -i '$i    error_page 404 =200 /index.html;' /etc/nginx/sites-enabled/main.conf`

**Comment by DaniellMesquita** [comment]
Basically explaining for dumbs, how it works?

**Comment by darkdrgn2k** [comment]
It basically has 2 lines

### Firewall
`iptables -t nat -I PREROUTING -i wlan-ap -p tcp --dport 80 -j DNAT --to-destination 10.0.0.1:80 `

This means ANY IPv4 packets that will be forwarded to another device, destined for TCP port 80, re-write its destination to the node itself  10.0.0.1:80  (Break down of the command at the bottom if your interested)

### nginx
`error_page 404 =200 /index.html`

Because our nginx server PROBABLY doesn't have the random path the person is asking for (because we are not the server the person wanted to access) they would get a 404 error (IE http://99.99.99.99/myfile.html  instead  you will get the content of http://10.0.0.1/myfile.html and that file does not exist)

This command takes all 404 errors and instead presents them with /index.html (home page)


# Break down of iptables command

`iptables`
Firewall for IPv4

`-t nat -I PREROUTING`
We are using the `NAT` module. Rules go in the `PREROUTING` table. Processing of this table happens when a packet is do to forwarded to another device but before it is routed.

`-i wlan-ap`
Packet coming in on the Access Point

`-p tcp`
TCP Protocol

`--dport 80`
Destination Port 80 (TCP port 80 is HTTP)

`-j DNAT`
Tells the firewall we will be doing a Destination NAT.  This means that we will be changing the destination of the packet. 

`--to-destination 10.0.0.1:80`
Re write the packet destination to the wlan-ap ip address

Then the packet goes through the routing process and ends up on the node's web server instead of being sent of into the wild.

**Comment by DaniellMesquita** [comment]
Then, it only accepts IPv6 traffic?

**Comment by DaniellMesquita** [comment]
How this script can be ported to Android?

**Comment by makeworld-the-better-one** [comment]
@DaniellMesquita IPv6 traffic is accepted, this way Yggdrasil and CJDNS addresses are not redirected. Porting to Android is far beyond the scope of this project, and I have no Android networking knowledge.

**Pull request merged**